[WIP]: add the release notes for v8.5.8 (#23522)
pingcap/docs: releases/release-8.5.8.md

---
title: TiDB 8.5.8 Release Notes
summary: Learn about the improvements and bug fixes in TiDB 8.5.8.
---

# TiDB 8.5.8 Release Notes

Release date: TBD

TiDB version: 8.5.8

Quick access: [Quick start](https://docs.pingcap.com/tidb/v8.5/quick-start-with-tidb) | [Production deployment](https://docs.pingcap.com/tidb/v8.5/production-deployment-using-tiup)

## Improvements

+ TiDB

    - Improve optimizer row count estimation for composite indexes that include virtual columns by falling back to index statistics when column statistics are unavailable, helping TiDB choose indexes more accurately [#69134](https://github.com/pingcap/tidb/issues/69134) @[qw4990](https://github.com/qw4990) <!-- component: planner --> <!-- pr: https://github.com/pingcap/tidb/pull/70327 -->

+ Tools

    + TiCDC

        - Improve TiCDC changefeed scan performance when delete events are ignored, reducing unnecessary DML decoding during historical catch-up for delete-heavy workloads [#5430](https://github.com/pingcap/ticdc/issues/5430) @[asddongmen](https://github.com/asddongmen) <!-- component: cdc --> <!-- pr: https://github.com/pingcap/ticdc/pull/5767 -->

>         - Support configuring TiCDC event collector batching by both event count and total bytes for changefeeds, so you can achieve more consistent batch sizes across different workloads and sinks [#3237](https://github.com/pingcap/ticdc/issues/3237) @[3AceShowHand](https://github.com/3AceShowHand) <!-- component: cdc --> <!-- pr: https://github.com/pingcap/ticdc/pull/5942 -->

Author: ```suggestion
```

Author: confirmed by @3AceShowHand, this configuration is not user visible, so it should be deleted from the release notes.

    - Fix the issue that non-privileged users can read other users' attributes through `INFORMATION_SCHEMA.USER_ATTRIBUTES` [#70277](https://github.com/pingcap/tidb/issues/70277) @[djshow832](https://github.com/djshow832) <!-- component: sql-infra --> <!-- pr: https://github.com/pingcap/tidb/pull/70317 -->
    - Fix the issue that recording OOM diagnostic goroutine profiles might prolong stop-the-world pauses and increase query latency when memory usage frequently exceeds the alarm ratio [#62080](https://github.com/pingcap/tidb/issues/62080) @[YangKeao](https://github.com/YangKeao) <!-- component: sql-infra --> <!-- pr: https://github.com/pingcap/tidb/pull/70228 -->
    - Fix the issue that `LOAD DATA LOCAL INFILE` in a pessimistic transaction might retry internally after a retryable lock conflict, desynchronize the client connection, and return an invalid sequence error instead of the original deadlock error [#69793](https://github.com/pingcap/tidb/issues/69793) @[lance6716](https://github.com/lance6716) <!-- component: transaction, sql-infra, execution --> <!-- pr: https://github.com/pingcap/tidb/pull/70194 -->

>     - Fix the issue that `ALTER TABLE ... REORGANIZE PARTITION` might rebuild global indexes without entries for rows in non-reorganized partitions that come after the reorganized partitions in partition order, causing queries using those indexes to miss rows and unique constraints not to be enforced for the affected values [#70023](https://github.com/pingcap/tidb/issues/70023) @[mjonss](https://github.com/mjonss) <!-- component: ddl --> <!-- pr: https://github.com/pingcap/tidb/pull/70479 -->

Author: ```suggestion
    - Fix the issue that `ALTER TABLE ... REORGANIZE PARTITION` might rebuild global indexes without entries for rows in non-reorganized partitions that come after the reorganized partitions in partition order, causing queries using those indexes to miss rows and potentially allowing duplicate index values to be inserted [#70023](https://github.com/pingcap/tidb/issues/70023) @[mjonss](https://github.com/mjonss) <!-- component: ddl --> <!-- pr: https://github.com/pingcap/tidb/pull/70479 -->
```

    - Fix the issue that TiKV performs unnecessary slow-log message formatting for fast Raftstore message batches, causing additional CPU overhead [#19861](https://github.com/tikv/tikv/issues/19861) @[pingyu](https://github.com/pingyu) <!-- component: tikv --> <!-- pr: https://github.com/tikv/tikv/pull/19910 -->
    - Fix the issue that BR log backup leaves stale GC safepoints after a log backup task is stopped, which might affect cleanup and safepoint management [#19832](https://github.com/tikv/tikv/issues/19832) @[Leavrth](https://github.com/Leavrth) <!-- component: br --> <!-- pr: https://github.com/tikv/tikv/pull/19911 -->

>     - Fix the issue that BR cannot update the SST download rate limit correctly when multiple restore tasks run concurrently, which might cause one task's limit change not to take effect [#19454](https://github.com/tikv/tikv/issues/19454) @[Leavrth](https://github.com/Leavrth) <!-- component: br --> <!-- pr: https://github.com/tikv/tikv/pull/19924 -->

Author: ```suggestion
```

    - Fix the issue that PD might establish outbound gRPC connections to caller-selected addresses supplied through `pd-forwarded-host` instead of restricting forwarding targets to the current PD leader's advertised client URLs [#11070](https://github.com/tikv/pd/issues/11070) @[rleungx](https://github.com/rleungx) <!-- component: pd --> <!-- pr: https://github.com/tikv/pd/pull/11091 -->
    - Fix the issue that the resource group client might permanently send `NaN` token requests when a newly created resource group controller races with the periodic state update [#11022](https://github.com/tikv/pd/issues/11022) @[JmPotato](https://github.com/JmPotato) <!-- component: pd --> <!-- pr: https://github.com/tikv/pd/pull/11028 -->

+ Tools

    + Backup & Restore (BR)

        - Fix the issue that BR log restore does not enforce the configured rate limit during point-in-time recovery, causing log apply download speed to exceed the limit [#63505](https://github.com/pingcap/tidb/issues/63505) @[Leavrth](https://github.com/Leavrth) <!-- component: br --> <!-- pr: https://github.com/pingcap/tidb/pull/69153 -->

>         - Fix the issue that BR point-in-time restore for tables with `AUTO_ID_CACHE=1` might cause duplicate-key errors on the first `INSERT` after restore [#69485](https://github.com/pingcap/tidb/issues/69485) @[vldmit](https://github.com/vldmit) <!-- component: br --> <!-- pr: https://github.com/pingcap/tidb/pull/70253 -->

Author: ```suggestion
        - Fix the issue that BR log restore does not enforce the configured rate limit during point-in-time recovery, causing log apply download speed to exceed the limit [#63505](https://github.com/pingcap/tidb/issues/63505) @[Leavrth](https://github.com/Leavrth) <!-- component: br --> <!-- pr: https://github.com/pingcap/tidb/pull/69153 -->
        - Fix the issue that BR point-in-time restore for tables with `AUTO_ID_CACHE=1` might cause duplicate-key errors on the first `INSERT` after restore [#69485](https://github.com/pingcap/tidb/issues/69485) @[vldmit](https://github.com/vldmit) <!-- component: br --> <!-- pr: https://github.com/pingcap/tidb/pull/70253 -->
        - Fix the issue that BR log backup leaves stale GC safepoints after a log backup task is stopped, which might affect cleanup and safepoint management [#19832](https://github.com/tikv/tikv/issues/19832) @[Leavrth](https://github.com/Leavrth) <!-- component: br --> <!-- pr: https://github.com/tikv/tikv/pull/19911 -->
        - Fix the issue that BR cannot update the SST download rate limit correctly when multiple restore tasks run concurrently, which might cause one task's limit change not to take effect [#19454](https://github.com/tikv/tikv/issues/19454) @[Leavrth](https://github.com/Leavrth) <!-- component: br --> <!-- pr: https://github.com/tikv/tikv/pull/19924 -->
```

---
title: TiDB 8.5.8 Release Notes
summary: Learn about the improvements and bug fixes in TiDB 8.5.8.
---

# TiDB 8.5.8 Release Notes

Release date: TBD

TiDB version: 8.5.8

Quick access: [Quick start](https://docs.pingcap.com/tidb/v8.5/quick-start-with-tidb) | [Production deployment](https://docs.pingcap.com/tidb/v8.5/production-deployment-using-tiup)

## Improvements

+ TiDB

    - Improve memory efficiency for joins and `UPDATE`/`DELETE` statements by reducing excessive initial chunk allocation, especially in high-concurrency workloads and when processing wide rows [#68545](https://github.com/pingcap/tidb/issues/68545) @[solotzg](https://github.com/solotzg) <!-- component: execution --> <!-- pr: https://github.com/pingcap/tidb/pull/69965 -->
    - Improve optimizer row count estimation for composite indexes that include virtual columns by falling back to index statistics when column statistics are unavailable, helping TiDB choose indexes more accurately [#69134](https://github.com/pingcap/tidb/issues/69134) @[qw4990](https://github.com/qw4990) <!-- component: planner --> <!-- pr: https://github.com/pingcap/tidb/pull/70327 -->

+ TiKV

>     - Reduce CPU overhead in TiKV raftstore peer and store FSM message handling by avoiding eager slow-log message formatting for fast message batches [#19861](https://github.com/tikv/tikv/issues/19861) @[pingyu](https://github.com/pingyu) <!-- component: tikv --> <!-- pr: https://github.com/tikv/tikv/pull/19910 -->

Author: ```suggestion
    - Reduce CPU overhead in TiKV Raftstore message handling by deferring slow-log formatting for fast message batches [#19861](https://github.com/tikv/tikv/issues/19861) @[pingyu](https://github.com/pingyu) <!-- component: tikv --> <!-- pr: https://github.com/tikv/tikv/pull/19910 -->
```

> 

Author: ```suggestion
```

+ Tools

    + TiCDC

        - Improve TiCDC changefeed scan performance when delete events are ignored, reducing unnecessary DML decoding during historical catch-up for delete-heavy workloads [#5430](https://github.com/pingcap/ticdc/issues/5430) @[asddongmen](https://github.com/asddongmen) <!-- component: cdc --> <!-- pr: https://github.com/pingcap/ticdc/pull/5767 -->
        - Support configuring TiCDC event collector batching by both event count and total bytes for changefeeds, so you can tune batch size more consistently across different workloads and sinks [#3237](https://github.com/pingcap/ticdc/issues/3237) @[3AceShowHand](https://github.com/3AceShowHand) <!-- component: cdc --> <!-- pr: https://github.com/pingcap/ticdc/pull/5942 -->
        - Improve TiCDC event service stability and throughput under memory pressure by introducing an adaptive scan window algorithm, reducing dispatcher starvation and reset events during DDL and sync point workloads [#4172](https://github.com/pingcap/ticdc/issues/4172) @[asddongmen](https://github.com/asddongmen) <!-- component: cdc --> <!-- pr: https://github.com/pingcap/ticdc/pull/5761 -->

## Bug fixes

+ TiDB

>     - Fix the issue that multi-schema changes containing `MODIFY COLUMN` on release-8.5 fall back to transactional backfill instead of using ingest or distributed backfill [#70136](https://github.com/pingcap/tidb/issues/70136) @[joechenrh](https://github.com/joechenrh) <!-- component: ddl --> <!-- pr: https://github.com/pingcap/tidb/pull/70188 -->

Author: ```suggestion
    - Fix the issue that multi-schema changes containing `MODIFY COLUMN` fall back to transactional backfill instead of using ingest or distributed backfill [#70136](https://github.com/pingcap/tidb/issues/70136) @[joechenrh](https://github.com/joechenrh) <!-- component: ddl --> <!-- pr: https://github.com/pingcap/tidb/pull/70188 -->
```

    - Fix the issue that TiDB might panic and terminate the user session when a query uses an `IN` subquery as an operand of a nested `NOT IN` expression [#64854](https://github.com/pingcap/tidb/issues/64854) @[hawkingrei](https://github.com/hawkingrei) <!-- component: planner --> <!-- pr: https://github.com/pingcap/tidb/pull/70259 -->
    - Fix the issue that otherwise cache-safe queries using `JSON_EXTRACT()` cannot use the prepared or non-prepared plan cache [#69522](https://github.com/pingcap/tidb/issues/69522) @[winoros](https://github.com/winoros) <!-- component: planner --> <!-- pr: https://github.com/pingcap/tidb/pull/70257 -->
    - Fix the issue that statistics for virtual generated columns added with `ALTER TABLE ADD COLUMN` might be initialized incorrectly, causing inaccurate `SHOW STATS_HISTOGRAMS` output and unnecessary loading of placeholder statistics [#69160](https://github.com/pingcap/tidb/issues/69160) @[qw4990](https://github.com/qw4990) <!-- component: planner --> <!-- pr: https://github.com/pingcap/tidb/pull/70325 -->

>     - Fix the issue that TiDB might consume unbounded connection memory when receiving repeated `COM_STMT_SEND_LONG_DATA` requests for prepared statements without executing or resetting them [#70349](https://github.com/pingcap/tidb/issues/70349) @[djshow832](https://github.com/djshow832) <!-- component: sql-infra --> <!-- pr: https://github.com/pingcap/tidb/pull/70369 -->

Author: ```suggestion
    - Fix the issue that `COM_STMT_SEND_LONG_DATA` parameter data could accumulate across multiple prepared statements in the same connection and cause unbounded connection memory usage; TiDB now accounts this parameter memory against the session memory quota [#70349](https://github.com/pingcap/tidb/issues/70349) @[djshow832](https://github.com/djshow832) <!-- component: sql-infra --> <!-- pr: https://github.com/pingcap/tidb/pull/70369 -->
```

>     - Fix the issue that TiDB might crash when parsing SQL statements or optimizer hints with excessively deep nested parentheses by rejecting them with a parser error before a stack overflow occurs [#70192](https://github.com/pingcap/tidb/issues/70192) @[Debra-He](https://github.com/Debra-He) <!-- component: sql-infra --> <!-- pr: https://github.com/pingcap/tidb/pull/70264 -->

Author: ```suggestion
    - Fix the issue that TiDB might crash when parsing SQL statements or optimizer hints with excessively deep nested parentheses [#70192](https://github.com/pingcap/tidb/issues/70192) @[Debra-He](https://github.com/Debra-He) <!-- component: sql-infra --> <!-- pr: https://github.com/pingcap/tidb/pull/70264 -->
```

    - Fix the issue that parsing TiDB runtime traces might fail during batched TSO requests because the PD client emits invalid nested runtime trace regions [#69743](https://github.com/pingcap/tidb/issues/69743) @[YangKeao](https://github.com/YangKeao) <!-- component: sql-infra --> <!-- pr: https://github.com/pingcap/tidb/pull/70310 -->
    - Fix the issue that `COM_CHANGE_USER` authentication failure might leave the connection in an inconsistent session state [#69691](https://github.com/pingcap/tidb/issues/69691) @[bb7133](https://github.com/bb7133) <!-- component: sql-infra --> <!-- pr: https://github.com/pingcap/tidb/pull/70229 -->

>     - Fix the issue that `COM_STMT_SEND_LONG_DATA` can accumulate unbounded prepared-statement parameter data and increase connection memory usage; TiDB now enforces the session `max_allowed_packet` limit and returns a packet-too-large error when the accumulated size exceeds the limit [#69693](https://github.com/pingcap/tidb/issues/69693) @[bb7133](https://github.com/bb7133) <!-- component: sql-infra --> <!-- pr: https://github.com/pingcap/tidb/pull/70231 -->

Author: ```suggestion
    - Fix the issue that `COM_STMT_SEND_LONG_DATA` could accumulate unbounded parameter data for a single prepared statement; TiDB now limits the accumulated size using the session `max_allowed_packet` setting and returns a packet-too-large error when the limit is exceeded [#69693](https://github.com/pingcap/tidb/issues/69693) @[bb7133](https://github.com/bb7133) <!-- component: sql-infra --> <!-- pr: https://github.com/pingcap/tidb/pull/70231 -->
```

    - Fix the issue that `LOAD DATA LOCAL INFILE` in a pessimistic transaction might retry internally after a retryable lock conflict, desynchronize the client connection, and return an invalid sequence error instead of the original deadlock error [#69793](https://github.com/pingcap/tidb/issues/69793) @[lance6716](https://github.com/lance6716) <!-- component: transaction, sql-infra, execution --> <!-- pr: https://github.com/pingcap/tidb/pull/70194 -->
    - Fix the issue that `ALTER TABLE ... REORGANIZE PARTITION` might rebuild global indexes without entries for rows in non-reorganized partitions located after the reorganized partitions, causing queries using those indexes to miss rows and unique constraints not to be enforced for the affected values [#70023](https://github.com/pingcap/tidb/issues/70023) @[mjonss](https://github.com/mjonss) <!-- component: ddl --> <!-- pr: https://github.com/pingcap/tidb/pull/70479 -->

+ TiKV

>     - Fix the issue that BR log backup leaves stale GC safepoints after a log backup task is stopped, which might affect cleanup and safepoint management [#19832](https://github.com/tikv/tikv/issues/19832) @[Leavrth](https://github.com/Leavrth) <!-- component: br --> <!-- pr: https://github.com/tikv/tikv/pull/19911 -->

Author: ```suggestion
    - Fix the issue that TiKV performs unnecessary slow-log message formatting for fast Raftstore message batches, causing additional CPU overhead [#19861](https://github.com/tikv/tikv/issues/19861) @[pingyu](https://github.com/pingyu) <!-- component: tikv --> <!-- pr: https://github.com/tikv/tikv/pull/19910 -->
    - Fix the issue that BR log backup leaves stale GC safepoints after a log backup task is stopped, which might affect cleanup and safepoint management [#19832](https://github.com/tikv/tikv/issues/19832) @[Leavrth](https://github.com/Leavrth) <!-- component: br --> <!-- pr: https://github.com/tikv/tikv/pull/19911 -->
```

    - Fix the issue that TiKV might generate inconsistent MVCC state during external SST ingestion when ingestion races with foreground writes, which might cause a panic in transaction status checks [#19891](https://github.com/tikv/tikv/issues/19891) @[gengliqi](https://github.com/gengliqi) <!-- component: tikv --> <!-- pr: https://github.com/tikv/tikv/pull/19916 -->
    - Fix the issue that TiKV cannot be built with `ENABLE_FIPS=1` in FIPS environments [#19743](https://github.com/tikv/tikv/issues/19743) @[LykxSassinator](https://github.com/LykxSassinator) <!-- component: tikv --> <!-- pr: https://github.com/tikv/tikv/pull/19753 --> <!-- pr: https://github.com/tikv/tikv/pull/19904 --> <!-- pr: https://github.com/tikv/tikv/pull/19754 -->
    - Fix the issue that enabling only background task resource control might cause a 5% to 10% performance regression in write-heavy workloads because TiKV unnecessarily switches the transaction scheduler to priority scheduling [#19858](https://github.com/tikv/tikv/issues/19858) @[glorv](https://github.com/glorv) <!-- component: tikv --> <!-- pr: https://github.com/tikv/tikv/pull/19895 -->

>     - Fix the issue that TiKV might crash with a null pointer dereference during TiDB Lightning data import in `tidb` mode [#18671](https://github.com/tikv/tikv/issues/18671) @[Dog-Du](https://github.com/Dog-Du) <!-- component: tikv --> <!-- pr: https://github.com/tikv/tikv/pull/19907 -->

Author: ```suggestion
    - Fix the issue that TiKV might crash during TiDB Lightning data import in `tidb` mode [#18671](https://github.com/tikv/tikv/issues/18671) @[Dog-Du](https://github.com/Dog-Du) <!-- component: tikv --> <!-- pr: https://github.com/tikv/tikv/pull/19907 -->
```

>     - Fix the issue that the TiKV In-Memory Engine region label watch might retry forever from a stale revision after an etcd compaction error is returned as a raw gRPC `UNKNOWN` status, and recover automatically by reloading all region labels and restarting the watch from a fresh revision [#19792](https://github.com/tikv/tikv/issues/19792) @[akashchakrabortymsc-cmd](https://github.com/akashchakrabortymsc-cmd) <!-- component: tikv --> <!-- pr: https://github.com/tikv/tikv/pull/19898 -->

Author: ```suggestion
    - Fix the issue that the TiKV In-Memory Engine might stop updating region labels after an etcd compaction error, causing the label watch to retry indefinitely [#19792](https://github.com/tikv/tikv/issues/19792) @[akashchakrabortymsc-cmd](https://github.com/akashchakrabortymsc-cmd) <!-- component: tikv --> <!-- pr: https://github.com/tikv/tikv/pull/19898 -->
```

    - Fix the issue that Region leader transfer might be blocked indefinitely when in-memory engine cache warmup is stuck and repeated transfer requests arrive before the ACK deadline [#19776](https://github.com/tikv/tikv/issues/19776) @[overvenus](https://github.com/overvenus) <!-- component: tikv --> <!-- pr: https://github.com/tikv/tikv/pull/19921 -->

>     - Fix the issue that TiKV applies unnecessary flow control during RocksDB base-level transitions because transient pending compaction bytes spikes are treated as sustained compaction pressure [#19667](https://github.com/tikv/tikv/issues/19667) @[hbisheng](https://github.com/hbisheng) <!-- component: tikv --> <!-- pr: https://github.com/tikv/tikv/pull/19828 -->

Author: ```suggestion
    - Fix the issue that TiKV might apply unnecessary write flow control during transient RocksDB compaction spikes [#19667](https://github.com/tikv/tikv/issues/19667) @[hbisheng](https://github.com/hbisheng) <!-- component: tikv --> <!-- pr: https://github.com/tikv/tikv/pull/19828 -->
```

+ PD

>     - Fix the issue that PD `/metric/query` and `/metric/query_range` might expose upstream response details or access invalid metric-storage targets by validating HTTP or HTTPS destinations, rejecting redirects, sanitizing proxy failures, and applying a 30-second timeout; malformed forwarded PD leader URLs with nested HTTP or HTTPS schemes are also rejected [#11081](https://github.com/tikv/pd/issues/11081) @[rleungx](https://github.com/rleungx) <!-- component: pd --> <!-- pr: https://github.com/tikv/pd/pull/11107 -->

Author: ```suggestion
    - Fix the issue that PD `/metric/query` and `/metric/query_range` might expose upstream response details or access invalid metric-storage targets @[rleungx](https://github.com/rleungx) <!-- component: pd --> <!-- pr: https://github.com/tikv/pd/pull/11107 -->
```

        - Fix the issue that changefeeds might disappear after upgrading TiCDC when existing changefeed ID metadata uses the legacy `namespace` field instead of `keyspace` in JSON [#4079](https://github.com/pingcap/ticdc/issues/4079) @[wlwilliamx](https://github.com/wlwilliamx) <!-- component: cdc --> <!-- pr: https://github.com/pingcap/ticdc/pull/5332 -->
        - Fix the issue that TiCDC checkpoint advancement gets stuck when replicating `CREATE TABLE ... LIKE ...` and the referenced source table is filtered out by the changefeed [#5150](https://github.com/pingcap/ticdc/issues/5150) @[lidezhu](https://github.com/lidezhu) <!-- component: cdc --> <!-- pr: https://github.com/pingcap/ticdc/pull/5951 -->
        - Fix the issue that TiCDC owner checkpoint timestamp and lag metrics remain stale after a changefeed is finished, stopped, removed, or moved to another owner [#5490](https://github.com/pingcap/ticdc/issues/5490) @[asddongmen](https://github.com/asddongmen) <!-- component: cdc --> <!-- pr: https://github.com/pingcap/ticdc/pull/5766 -->

>         - Fix the issue that a changefeed can stall after TiCDC failover or dispatcher reset in the new architecture [#5553](https://github.com/pingcap/ticdc/issues/5553) @[asddongmen](https://github.com/asddongmen) <!-- component: cdc --> <!-- pr: https://github.com/pingcap/ticdc/pull/5769 -->

Author: ```suggestion
        - Fix the issue that a changefeed can stall after TiCDC failover or dispatcher reset in the TiCDC new architecture [#5553](https://github.com/pingcap/ticdc/issues/5553) @[asddongmen](https://github.com/asddongmen) <!-- component: cdc --> <!-- pr: https://github.com/pingcap/ticdc/pull/5769 -->
```

>         - Fix the issue that a removing TiCDC maintainer might still reschedule or recreate dispatchers after shutdown handoff starts [#4827](https://github.com/pingcap/ticdc/issues/4827) @[wlwilliamx](https://github.com/wlwilliamx) <!-- component: cdc --> <!-- pr: https://github.com/pingcap/ticdc/pull/5403 -->

Author: ```suggestion
        - Fix the issue that a TiCDC maintainer being removed might still reschedule or recreate dispatchers after shutdown handoff starts [#4827](https://github.com/pingcap/ticdc/issues/4827) @[wlwilliamx](https://github.com/wlwilliamx) <!-- component: cdc --> <!-- pr: https://github.com/pingcap/ticdc/pull/5403 -->
```

        - Fix the issue that TiCDC stale-lock resolving might fail for async-commit residual locks because the `ScanLock` target timestamp can advance TiKV local MaxTS beyond the latest PD TSO [#5418](https://github.com/pingcap/ticdc/issues/5418) @[tenfyzhong](https://github.com/tenfyzhong) <!-- component: cdc --> <!-- pr: https://github.com/pingcap/ticdc/pull/5753 -->

>         - Fix the issue that next-gen `cdc cli` might show an empty `Error:` output when `--keyspace` or `-k` is missing, and now provides actionable guidance for missing or non-existent keyspaces [#2669](https://github.com/pingcap/ticdc/issues/2669) @[tenfyzhong](https://github.com/tenfyzhong) <!-- component: cdc --> <!-- pr: https://github.com/pingcap/ticdc/pull/5802 -->

Author: ```suggestion
        - Fix the issue that next-gen `cdc cli` might show an empty `Error:` output when `--keyspace` or `-k` is missing, and now provides actionable guidance for missing or non-existent keyspaces [#2669](https://github.com/pingcap/ticdc/issues/2669) @[tenfyzhong](https://github.com/tenfyzhong) <!-- component: cdc --> <!-- pr: https://github.com/pingcap/ticdc/pull/5802 -->
```

Author: reason: remove this release note, as users cannot access this feature directly and won't encouter this issue

---
title: TiDB 8.5.8 Release Notes
summary: Learn about the improvements and bug fixes in TiDB 8.5.8.
---

# TiDB 8.5.8 Release Notes

Release date: TBD

TiDB version: 8.5.8

Quick access: [Quick start](https://docs.pingcap.com/tidb/v8.5/quick-start-with-tidb) | [Production deployment](https://docs.pingcap.com/tidb/v8.5/production-deployment-using-tiup)

## Improvements

+ TiDB

    - Improve memory efficiency for joins and `UPDATE`/`DELETE` statements by reducing excessive initial chunk allocation, especially in high-concurrency workloads and when processing wide rows [#68545](https://github.com/pingcap/tidb/issues/68545) @[solotzg](https://github.com/solotzg) <!-- component: execution --> <!-- pr: https://github.com/pingcap/tidb/pull/69965 -->
    - Improve optimizer row count estimation for composite indexes that include virtual columns by falling back to index statistics when column statistics are unavailable, helping TiDB choose indexes more accurately [#69134](https://github.com/pingcap/tidb/issues/69134) @[qw4990](https://github.com/qw4990) <!-- component: planner --> <!-- pr: https://github.com/pingcap/tidb/pull/70327 -->

+ TiKV

>     - Add the optional TiKV configuration parameter `server.end-point-max-bg-concurrency` to limit background `ANALYZE` task concurrency independently of ordinary coprocessor requests, preventing `ANALYZE` from occupying shared coprocessor concurrency and reducing performance degradation and request latency under high concurrency [#19623](https://github.com/tikv/tikv/issues/19623) [#19969](https://github.com/tikv/tikv/issues/19969) @[LykxSassinator](https://github.com/LykxSassinator) <!-- component: tikv --> <!-- pr: https://github.com/tikv/tikv/pull/19973 --> <!-- pr: https://github.com/tikv/tikv/pull/19926 -->

Reviewer: Since this newly introduced configuration is hidden for the user, this release note is redundant.
```suggestion
```

    - Fix the issue that normal users can read other users' attributes through `INFORMATION_SCHEMA.USER_ATTRIBUTES` [#70277](https://github.com/pingcap/tidb/issues/70277) @[djshow832](https://github.com/djshow832) <!-- component: sql-infra --> <!-- pr: https://github.com/pingcap/tidb/pull/70317 -->
    - Fix the issue that recording OOM diagnostic goroutine profiles might increase stop-the-world latency and affect query latency when memory usage frequently exceeds the alarm ratio [#62080](https://github.com/pingcap/tidb/issues/62080) @[YangKeao](https://github.com/YangKeao) <!-- component: sql-infra --> <!-- pr: https://github.com/pingcap/tidb/pull/70228 -->
    - Fix the issue that `LOAD DATA LOCAL INFILE` in a pessimistic transaction might retry internally after a retryable lock conflict, desynchronize the client connection, and return an invalid sequence error instead of the original deadlock error [#69793](https://github.com/pingcap/tidb/issues/69793) @[lance6716](https://github.com/lance6716) <!-- component: transaction, sql-infra, execution --> <!-- pr: https://github.com/pingcap/tidb/pull/70194 -->

> - Fix the issue that `ALTER TABLE ... REORGANIZE PARTITION` might rebuild global indexes without rows from non-reorganized partitions ordered after the reorganized partitions, causing queries using those global indexes to return missing rows and unique constraints not to be enforced for those rows [#70023](https://github.com/pingcap/tidb/issues/70023) @[mjonss](https://github.com/mjonss) <!-- component: ddl --> <!-- pr: https://github.com/pingcap/tidb/pull/70479 -->

Author: ```suggestion
    - Fix the issue that `ALTER TABLE ... REORGANIZE PARTITION` might rebuild global indexes without entries for rows in non-reorganized partitions that follow the reorganized partitions, causing queries using those indexes to miss rows and unique constraints not to be enforced for the affected values [#70023](https://github.com/pingcap/tidb/issues/70023) @[mjonss](https://github.com/mjonss) <!-- component: ddl --> <!-- pr: https://github.com/pingcap/tidb/pull/70479 -->
```

    - Reduce CPU overhead in TiKV raftstore peer and store FSM message handling by avoiding eager slow-log message formatting for fast message batches [#19861](https://github.com/tikv/tikv/issues/19861) @[pingyu](https://github.com/pingyu) <!-- component: tikv --> <!-- pr: https://github.com/tikv/tikv/pull/19910 -->

+ Tools

    + TiCDC

        - Improve TiCDC changefeed scan performance when delete events are ignored, reducing unnecessary DML decoding during historical catch-up for delete-heavy workloads [#5430](https://github.com/pingcap/ticdc/issues/5430) @[asddongmen](https://github.com/asddongmen) <!-- component: cdc --> <!-- pr: https://github.com/pingcap/ticdc/pull/5767 -->

>         - Improve TiCDC Kafka sink behavior by separating the changefeed `max-message-bytes` batching threshold from Kafka's actual message size limit, so large row events that are within the Kafka topic or broker limit can be sent without changing the changefeed configuration [#1405](https://github.com/pingcap/ticdc/issues/1405) @[3AceShowHand](https://github.com/3AceShowHand) <!-- component: cdc --> <!-- pr: https://github.com/pingcap/ticdc/pull/5772 -->

Reviewer 2: ```suggestion
```

        - Support configuring TiCDC event collector batching by both event count and total bytes for changefeeds, so you can tune batch size more consistently across different workloads and sinks [#3237](https://github.com/pingcap/ticdc/issues/3237) @[3AceShowHand](https://github.com/3AceShowHand) <!-- component: cdc --> <!-- pr: https://github.com/pingcap/ticdc/pull/5942 -->
        - Improve TiCDC event service stability and throughput under memory pressure by introducing an adaptive scan window algorithm, reducing dispatcher starvation and reset events during DDL and sync point workloads [#4172](https://github.com/pingcap/ticdc/issues/4172) @[asddongmen](https://github.com/asddongmen) <!-- component: cdc --> <!-- pr: https://github.com/pingcap/ticdc/pull/5761 -->

>         - Add the `ignore-update-only-columns` event filter option for TiCDC Kafka sinks to skip `UPDATE` events when only the specified columns change, reducing unnecessary downstream traffic [#5269](https://github.com/pingcap/ticdc/issues/5269) @[lidezhu](https://github.com/lidezhu) <!-- component: cdc --> <!-- pr: https://github.com/pingcap/ticdc/pull/5738 -->

Reviewer 2: ```suggestion
```

    - Fix the issue that normal users can read other users' attributes through `INFORMATION_SCHEMA.USER_ATTRIBUTES` [#70277](https://github.com/pingcap/tidb/issues/70277) @[djshow832](https://github.com/djshow832) <!-- component: sql-infra --> <!-- pr: https://github.com/pingcap/tidb/pull/70317 -->
    - Fix the issue that recording OOM diagnostic goroutine profiles might increase stop-the-world latency and affect query latency when memory usage frequently exceeds the alarm ratio [#62080](https://github.com/pingcap/tidb/issues/62080) @[YangKeao](https://github.com/YangKeao) <!-- component: sql-infra --> <!-- pr: https://github.com/pingcap/tidb/pull/70228 -->
    - Fix the issue that `LOAD DATA LOCAL INFILE` in a pessimistic transaction might retry internally after a retryable lock conflict, desynchronize the client connection, and return an invalid sequence error instead of the original deadlock error [#69793](https://github.com/pingcap/tidb/issues/69793) @[lance6716](https://github.com/lance6716) <!-- component: transaction, sql-infra, execution --> <!-- pr: https://github.com/pingcap/tidb/pull/70194 -->

> 

Reviewer 3: ```suggestion
- Fix the issue that `ALTER TABLE ... REORGANIZE PARTITION` might rebuild global indexes without rows from non-reorganized partitions ordered after the reorganized partitions, causing queries using those global indexes to return missing rows and unique constraints not to be enforced for those rows [#70023](https://github.com/pingcap/tidb/issues/70023) @[mjonss](https://github.com/mjonss) <!-- component: ddl --> <!-- pr: https://github.com/pingcap/tidb/pull/70479 -->

```

Reviewer 3: This one is missing because it's merged after code freeze.

    + TiDB Lightning

        - Fix the issue that `IMPORT INTO` might report success with inconsistent indexes after a transient conflict-deletion commit error [#69792](https://github.com/pingcap/tidb/issues/69792) @[D3Hunter](https://github.com/D3Hunter) <!-- component: lightning --> <!-- pr: https://github.com/pingcap/tidb/pull/70254 -->
        - Fix the issue that `IMPORT INTO` might fail with the `lock held by current process` error after an import error or retry when local engine files are not cleaned up correctly [#65645](https://github.com/pingcap/tidb/issues/65645) @[D3Hunter](https://github.com/D3Hunter) <!-- component: lightning --> <!-- pr: https://github.com/pingcap/tidb/pull/69478 -->
        - Fix the issue that `IMPORT INTO` silently writes incorrect values for repeated dictionary-encoded Parquet `DECIMAL` columns [#70365](https://github.com/pingcap/tidb/issues/70365) @[joechenrh](https://github.com/joechenrh) <!-- component: lightning --> <!-- pr: https://github.com/pingcap/tidb/pull/70461 -->

>         - Fix the issue that `IMPORT INTO` or related local backend ingest tasks might hang when job generation is canceled [#69240](https://github.com/pingcap/tidb/issues/69240) @[D3Hunter](https://github.com/D3Hunter) <!-- component: lightning, dxf, lightning --> <!-- pr: https://github.com/pingcap/tidb/pull/70234 -->

Reviewer 3: ```suggestion
        - Fix the issue that `IMPORT INTO` or related local backend ingest tasks might hang when job generation is canceled [#69240](https://github.com/pingcap/tidb/issues/69240) @[D3Hunter](https://github.com/D3Hunter) <!-- component: lightning, dxf --> <!-- pr: https://github.com/pingcap/tidb/pull/70234 -->
```
